File: backend_android/src/main/java/com/darkraha/backend_android/components/images/gif/GifDrawable.java
package com.darkraha.backend_android.components.images.gif;

import android.graphics.*;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.Drawable;
import android.os.SystemClock;
import com.darkraha.backend_android.components.endecode.GifDecoder;

/**
 * Created by Verma Rahul
 * <p/>
 * Tested on width=match_parent , height = wrap_content
 * </p>
 * Use properly stop and start methoths, as example after assign to ImageView in RecyclingView
 * invoke stop(); start();
 */
public class GifDrawable extends Drawable implements Animatable, Runnable {
    private static final String TAG = "GifDrawable";
    protected Bitmap curBitmap;


    //protected int mCounter = 0;
    protected boolean alwaysAnimate;
    protected boolean mRunning;
    protected long mLastTime;
    protected Paint mPaint = new Paint();
    protected Matrix mMatrix = new Matrix();
    protected GifDecoder gifDecoder;
    protected int defaultDelay = 200;
    protected int mCountFrames;

    public GifDrawable(GifDecoder gifDecoder) {
        this.gifDecoder = gifDecoder;
        mPaint.setAntiAlias(true);
        mCountFrames = gifDecoder.getFrameCount();
        curBitmap = gifDecoder.getNextFrame();
    }

    public GifDecoder getGifDecoder() {
        return gifDecoder;
    }

    protected void setFirstFrame() {
        mLastTime = SystemClock.uptimeMillis();
        scheduleSelf(this, 0);
        invalidateSelf();
    }

    /**
     * Set default delay value when gif delay for frame == 0.
     *
     * @param value in ms
     */
    public void setDefaultDelay(int value) {
        defaultDelay = value;
    }

    public int getSize() {
        return gifDecoder.getSize();
    }

    //==============================================================


    @Override
    public int getIntrinsicWidth() {
        return gifDecoder.getWidth();
    }

    @Override
    public int getIntrinsicHeight() {
        return gifDecoder.getHeight();
    }

//    @Override
//    protected void onBoundsChange(Rect bounds) {
//
//        mMatrix.setRectToRect(new RectF(0, 0, curBitmap.getWidth(), curBitmap.getHeight()),
//                new RectF(bounds),
//                Matrix.ScaleToFit.CENTER);
//    }

    @Override
    public void draw(Canvas canvas) {

        if (curBitmap != null) {
            canvas.drawBitmap(curBitmap, mMatrix, mPaint);
        } else {
            setFirstFrame();
        }
    }

    @Override
    public void setAlpha(int alpha) {
        mPaint.setAlpha(alpha);
    }

    @Override
    public void setColorFilter(ColorFilter colorFilter) {
        mPaint.setColorFilter(colorFilter);
    }

    @Override
    public int getOpacity() {
        return 0; //PixelFormat.OPAQUE;
    }


    //==============================================================
    @Override
    public void start() {

        if (isRunning() || mCountFrames == 0)
            return;
        mRunning = true;
        mLastTime = SystemClock.uptimeMillis();
        scheduleSelf(this, 0);
        invalidateSelf();
    }

    @Override
    public void stop() {

        if (!alwaysAnimate) {
            mRunning = false;
            unscheduleSelf(this);
            invalidateSelf();
        }
    }

    @Override
    public boolean isRunning() {
        return mRunning;
    }


    public void setAlwaysAnimate(boolean v) {
        alwaysAnimate = v;
    }

    //==============================================================
    @Override
    public void run() {

        mLastTime = SystemClock.uptimeMillis();

        curBitmap = gifDecoder.getNextFrame();
        gifDecoder.advance();

        int delay = gifDecoder.getNextDelay();
        if (delay == 0) {
            delay = defaultDelay;
        }

        // check is running necessary for first frame
        if (isRunning()) {
            scheduleSelf(this, mLastTime + delay);
            invalidateSelf();
        }
    }
}

File: backend_android/src/main/java/com/darkraha/backend_android/components/images/gif/GifDrawable.kt
package com.darkraha.backend_android.components.images.gif

import android.graphics.*
import android.graphics.drawable.Animatable
import android.graphics.drawable.Drawable
import android.os.SystemClock
import com.darkraha.backend.Backend
import com.darkraha.backend_android.components.endecode.GifDecoder

/**
 * Created by Verma Rahul
 *
 *
 * Tested on width=match_parent , height = wrap_content
 *
 * Use properly stop and start methoths, as example after assign to ImageView in RecyclingView
 * invoke stop(); start();
 */
class GifDrawable(gifDecoder: GifDecoder) : Drawable(), Animatable, Runnable {
    protected var curBitmap: Bitmap? = null


    //protected int mCounter = 0;
    var alwaysAnimate: Boolean = false
    protected var mRunning: Boolean = false
    protected var mLastTime: Long = 0
    val mPaint = Paint()
    protected var mMatrix = Matrix()
    var gifDecoder: GifDecoder
        protected set
    var defaultDelay = 200
    var mCountFrames: Int = 0
        protected set

    /**
     * How long show frame image in animated mode.
     */
    val frameDelay: Int
        get() = if (gifDecoder.nextDelay == 0) defaultDelay else gifDecoder.nextDelay


    val size: Int
        get() = gifDecoder.size

    init {
        this.gifDecoder = gifDecoder
        mPaint.isAntiAlias = true
        mCountFrames = gifDecoder.frameCount
        curBitmap = gifDecoder.nextFrame
    }

    protected fun setFirstFrame() {
        mLastTime = SystemClock.uptimeMillis()
        scheduleSelf(this, 0)
        invalidateSelf()
    }


    //==============================================================


    override fun getIntrinsicWidth(): Int {
        return gifDecoder.width
    }

    override fun getIntrinsicHeight(): Int {
        return gifDecoder.height
    }

    override fun draw(canvas: Canvas) {

        if (curBitmap != null) {
            canvas.drawBitmap(curBitmap!!, mMatrix, mPaint)
        } else {
            setFirstFrame()
        }
    }

    override fun setAlpha(alpha: Int) {
        mPaint.alpha = alpha
    }

    override fun setColorFilter(colorFilter: ColorFilter?) {
        mPaint.colorFilter = colorFilter
    }

    override fun getOpacity(): Int {
        return 0 //PixelFormat.OPAQUE;
    }


    //==============================================================
    override fun start() {

        if (isRunning || mCountFrames == 0)
            return
        mRunning = true
        mLastTime = SystemClock.uptimeMillis()
        scheduleSelf(this, 0)
        invalidateSelf()
    }

    override fun stop() {

        if (!alwaysAnimate) {
            mRunning = false
            unscheduleSelf(this)
            invalidateSelf()
        }
    }

    override fun isRunning(): Boolean {
        return mRunning
    }


    //==============================================================
    override fun run() {
//        if (!isRunning) {
//            return
//        }
        mLastTime = SystemClock.uptimeMillis()

        Backend.sharedInstance.exeAsync({
            curBitmap = gifDecoder.nextFrame
            gifDecoder.advance()
        }, {
            sheduleNextFrame()
        })
    }

    fun sheduleNextFrame() {
        if (isRunning) {
            scheduleSelf(this, mLastTime + frameDelay)
            println("GifDrawable on run")
            invalidateSelf()
        }
    }

    companion object {
        private val TAG = "GifDrawable"
    }
}
